homeassistant/components/modbus.py

```diff
@@ -175,10 +228,24 @@ def connect(self):
         """Connect client."""
         with self._lock:
             self._client.connect()
+            if isinstance(self._rs485_mode, dict):
+                from serial.rs485 import RS485Settings
+                rs485_mode = RS485Settings(
+                    delay_before_tx=(
+                        self._rs485_mode[CONF_DELAY_BEFORE_TX]),
+                    delay_before_rx=(
+                        self._rs485_mode[CONF_DELAY_BEFORE_RX]),
+                    rts_level_for_tx=(
+                        self._rs485_mode[CONF_RTS_LEVEL_FOR_TX]),
+                    rts_level_for_rx=(
+                        self._rs485_mode[CONF_RTS_LEVEL_FOR_RX]),
+                    loopback=False)
+                self._client.socket.rs485_mode = rs485_mode
 
     def read_coils(self, unit, address, count):
         """Read coils."""
         with self._lock:
+            time.sleep(self._delay_between_queries)
```

Comment:
You should not use sleep inside home-assistant. Use our helpers to task a function for later execution

Reply:
Hi! time.sleep is used in many  components (ex. recorder, cover/rpi_gpio, cover/velbus, lock/verisure, lirc, nuimo_controller, sensor/fritzbox_callmonitor... and others....). Anyway 
I did not find an equivalent helper. Could you suggest something?

Reply:
You could try and use `track_point_in_time` in helpers/event.py.

Reply:
track_point_in_time take a function as callback and call it later. Not exacly what I need. To use this helper I have to insert another lock that suspend the function and schedule a call that unlock it.

Reply:
You can eg add a boolean argument suspend to each function where you wanted to sleep. Then if the boolean is `True` when entering the function you use the helper to schedule another call to the same function but with the boolean as `False`. When the boolean is false, the function just continues past the check without suspension.

Reply:
You are right, but this approach don't solve the Main problem. In this way I ensure that one of the modbus function can't be called too often (like throttle). What we need is different. In modbus over rs485 we need a period of silence between queries. The protocol define a silence time and this is handled by pymodbus. Many modbus slaves need more silence time. For this reason amost all industrials HMIs have this parameter in configuration. 

...so... We have to ensure a pause between any function in modbus component. I can schedule a read registers 50 ms later and a write register 50 ms later and this give me a pause of 50ms between two write register and two read register, but this don't ensure a pause of 50ms between one read registers and one write register. 

....I hope that my poor English explained the case... :)

Reply:
I think you can solve this with a queue and `track_time_interval`. This helper will call a callback at every time timedelta interval. The callback should be to get a job from the queue and perform the job. The job should be a callback with arguments that performs the action and does something with the result from the action. Eg `read_coils` is the action that returns a result.

Reply:
This Mean rewrite all modbus.... :'(  and there are another problem: define a Fix interval is not optimal.  I have to know how long take each function and change the time interval every time.

Reply:
Ok, I think the best solution using threads is to create a new persistent thread for the mod business hub instance, that gets jobs from a queue. It's only a problem to sleep in the threads of the worker pool, not for other threads. So if you create a new persistent thread for the hub, you can sleep all you want in that thread. You can also define the sleep time with arguments to the jobs.

This is my suggestion. Maybe @pvizeli has a better idea?

Reply:
I've no idea how to do that but I will have a look on it. This can be the opportunity to support multiple modbus hub instances.

Reply:
@MartinHjelmare do you know if there are another hub that work as you described?

Reply:
Not a hub, but you can look at pymysensors library at github and its serial gateway module. The serial gateway class inherits from `threading.Thread` and polls a queue for work to send to the serial connection. Anything coming off the serial connection is also put on the queue. It could probably be improved with two different threads for reading writing to the serial connection as that's duplex.

Reply:
You are only allowed to use `asyncio.sleep`. Not `time.sleep`. For that, you need to be inside an async context and run your methods on client inside the executor.

Something like this (untested):

```python
@asyncio.coroutine
def async_read_input_register(self, unit, address, count):
    with (yield from self._paho_lock):
        yield from asyncio.sleep(self._delay)
        result = yield from self.hass.async_add_job(self._client.read_input_registers, address, count)
        return result
```

Note that it seems a bit odd to always sleep. Instead you probably want to track when the last request was done and sleep only if necessary.

Reply:
Ok I have to try. 
So I think to:
- reduce this pr to cover only hardware rs485 adapters 
- open a issues (if modbus.py is placed into custom_components there are an exception and I don't know exacly why)
- make another pr after tests

Reply:
@balloob and @MartinHjelmare I have a question.
Now modbus hub calls directly pymodbus function. These function are blocking. If I set a 30s timeout, and I call a function addressed to a non responding device, that function will return after 30s. As I understood it is almost the same as place a `timer.sleep`.
So what's the correct way to handle this situation?  Rewrite all in async style or have a different thread that handle the communication?

Reply:
Yeah I think you either have to rewrite pymodbus to use asyncio or use a persistent thread with a queue as I described above.

Reply:
I like thread and queue solution but implementation method stiIl not enough clear for me. But...  Is it a problem if I have a function that returns after 30s inside an async call?

Reply:
I think that would be a problem. Even within a worker pool thread, I don't think we want to occupy it for that long.

Reply:
It's a extreme case, but it could happpens. Default pymodbus timeout is 3s, but it is configurable.